File: frontend/src/types/roadmap.ts
export enum DayStatus {
    LOCKED = "LOCKED",
    UNLOCKED = "UNLOCKED",
    COMPLETED = "COMPLETED",
    SKIPPED = "SKIPPED"
}

export interface DayContent {
    day_number: number;
    topic: string;
    interaction_type: string;
    prompt: string;
    requirements?: Record<string, unknown>;
}

export interface DailyProgress {
    id: number;
    user_id: number;
    day_number: number;
    status: DayStatus;
    content_data?: Record<string, unknown>;
    completed_at?: string;
    reward_earned?: number;
}

export interface DailyProgressUpdate {
    content_data: Record<string, unknown>;
}

export interface RoadmapHistory {
    id: number;
    user_id: number;
    snapshot_data: Record<string, any>;
    is_premium: boolean;
    created_at: string;
}

File: backend/app/schemas/roadmap.py
from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel
from backend.app.models.roadmap import DayStatus

class DailyProgressBase(BaseModel):
    day_number: int
    status: DayStatus
    content_data: Optional[Any] = None

class DailyProgressCreate(DailyProgressBase):
    pass

class DailyProgressUpdate(BaseModel):
    content_data: Any
    status: Optional[DayStatus] = None

class RoadmapHistoryBase(BaseModel):
    snapshot_data: dict
    is_premium: bool = False

class RoadmapHistoryCreate(RoadmapHistoryBase):
    user_id: int

class RoadmapHistory(RoadmapHistoryBase):
    id: int
    user_id: int
    created_at: datetime

    class Config:
        from_attributes = True

class DailyProgress(DailyProgressBase):
    id: int
    user_id: int
    completed_at: Optional[datetime] = None
    reward_earned: Optional[int] = 0

    class Config:
        from_attributes = True

class DayContent(BaseModel):
    day_number: int
    topic: str
    interaction_type: str # "Journaling", "Checklist", "Scoring", "Selection"
    prompt: str
    requirements: Optional[Dict[str, Any]] = None # e.g. min_length: 50
